File: js/problem52.js
/**
 * @param {number} n
 * @return {number}
 */

let res = 0;

const initializeBoard = function(n) {
    let board = [];
    let row = [];
    
    for (let i = 0; i < n; i++) {
        row.push('.');
    }
    
    for (let i = 0; i < n; i++) {
        board.push(row.slice());
    }
    
    return board;
};


const canPlace = function(n, board, row, col) {
    for (let i = 0; i < col; i++) {
        if (board[row][i] === 'Q') {
            return false;
        }
    }
    
    for (let i = row - 1, j = col - 1; (i >= 0) && (j >= 0); i--, j--) {
        if (board[i][j] === 'Q') {
            return false;
        }
    }
    
    for (let i = row + 1, j = col - 1; (i < n) && (j >= 0); i++, j--) {
        if (board[i][j] === 'Q') {
            return false;
        } 
    }
    
    return true;
};

const placeQueen = function(n, board, col) {    
    if (col === n) {
        res++;
        return;
    }
    
    for (let row = 0; row < n; row++) {
        if (canPlace(n, board, row, col)) {
            board[row][col] = 'Q';
            placeQueen(n, board, col + 1);
            board[row][col] = '.';
        }
    }
};

var totalNQueens = function(n) {
    if (n === 1) {
        return 1;
    } else if (n <= 2) {
        return 0;
    }
    
    res = 0;
    
    let board = initializeBoard(n);
    placeQueen(n, board, 0);
    
    return res;
    
}

File: leetcodeJava/src/main/java/Problem52.java
// 52. N-Queens II

class Solution {
    
    private boolean canPlace(char[][] board, int row, int column) {
        for (int j = 0; j < column; j++) {
            if (board[row][j] == 'Q') {
                return false;
            }
        }
        for (int i = row - 1, j = column -1; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }
        for (int i = row + 1, j = column - 1; i < board.length && j >= 0; i++, j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }
        return true;
    }
    
    private int placeQueen(char[][] board, int column) {
        if (column == board.length) {
            return 1;
        }
        int res = 0;
        for (int i = 0; i < board.length; i++) {
            if (this.canPlace(board, i, column)) {
                board[i][column] = 'Q';
                res += this.placeQueen(board, column + 1);
                board[i][column] = '\u0000';
            }
        }
        return res;
    }
    
    public int totalNQueens(int n) {
        if (n == 1) {
            return 1;
        } else if (n < 4) {
            return 0;
        }
        char[][] board = new char[n][n];
        return this.placeQueen(board, 0);
    }
}